File: frontend-miniapp/scripts/generate_icons.py
"""Generate tabBar icons for WeChat mini-program."""
import struct
import zlib
import os

def create_png(width, height, pixels):
    """Create a valid PNG file from RGBA pixel data."""
    def chunk(chunk_type, data):
        c = chunk_type + data
        crc = struct.pack(">I", zlib.crc32(c) & 0xFFFFFFFF)
        return struct.pack(">I", len(data)) + c + crc

    # PNG signature
    sig = b"\x89PNG\r\n\x1a\n"

    # IHDR
    ihdr_data = struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0)
    ihdr = chunk(b"IHDR", ihdr_data)

    # IDAT - filter byte 0 for each row, then RGBA pixels
    raw = b""
    for y in range(height):
        raw += b"\x00"  # filter: None
        for x in range(width):
            raw += pixels[y * width + x]

    idat = chunk(b"IDAT", zlib.compress(raw))

    # IEND
    iend = chunk(b"IEND", b"")

    return sig + ihdr + idat + iend

def rgba(r, g, b, a=255):
    return struct.pack("BBBB", r, g, b, a)

def draw_circle(cx, cy, r, color, size=81):
    """Draw a filled circle on a transparent buffer."""
    pixels = [rgba(0, 0, 0, 0) for _ in range(size * size)]
    for y in range(size):
        for x in range(size):
            dx, dy = x - cx, y - cy
            if dx*dx + dy*dy <= r*r:
                pixels[y * size + x] = color
    return pixels

def draw_house(color, size=81):
    """Simple house shape."""
    pixels = [rgba(0, 0, 0, 0) for _ in range(size * size)]
    # Body
    for y in range(30, 75):
        for x in range(15, 66):
            pixels[y * size + x] = color
    # Roof
    for y in range(12, 35):
        roof_width = int((y - 12) * 50 / 23 + 10)
        for x in range(40 - roof_width, 41 + roof_width):
            if 0 <= x < size:
                pixels[y * size + x] = color
    # Door
    door_color = rgba(255, 255, 255, 255)
    for y in range(48, 75):
        for x in range(32, 49):
            pixels[y * size + x] = door_color
    return pixels

def draw_book(color, size=81):
    """Simple book shape for academic."""
    pixels = [rgba(0, 0, 0, 0) for _ in range(size * size)]
    # Book body
    for y in range(20, 65):
        for x in range(20, 61):
            pixels[y * size + x] = color
    # Spine line
    spine = rgba(255, 255, 255, 200)
    for y in range(20, 65):
        for x in range(38, 43):
            pixels[y * size + x] = spine
    # Lines on left page
    line_c = rgba(255, 255, 255, 150)
    for lx in range(24, 36):
        for ly in [28, 36, 44, 52]:
            pixels[ly * size + lx] = line_c
    # Lines on right page
    for lx in range(45, 57):
        for ly in [28, 36, 44, 52]:
            pixels[ly * size + lx] = line_c
    return pixels

def draw_document(color, size=81):
    """Simple document shape for resume."""
    pixels = [rgba(0, 0, 0, 0) for _ in range(size * size)]
    # Paper body
    for y in range(15, 70):
        for x in range(18, 63):
            pixels[y * size + x] = color
    # Fold corner
    for y in range(15, 30):
        for x in range(48, 63):
            if x - 48 >= (30 - y):
                pixels[y * size + x] = rgba(0, 0, 0, 0)
    # Lines
    line_c = rgba(255, 255, 255, 180)
    line_y = [28, 36, 44, 52, 60]
    for ly in line_y:
        for lx in range(26, 56):
            pixels[ly * size + lx] = line_c
    return pixels

def draw_person(color, size=81):
    """Simple person shape for profile."""
    pixels = [rgba(0, 0, 0, 0) for _ in range(size * size)]
    # Head (circle)
    for y in range(12, 38):
        for x in range(25, 56):
            dx, dy = x - 40.5, y - 25
            if dx*dx/(13*13) + dy*dy/(13*13) <= 1:
                pixels[y * size + x] = color
    # Body (trapezoid)
    for y in range(40, 72):
        body_width = 8 + (y - 40) * 6 // 16
        for x in range(40 - body_width, 41 + body_width):
            if 0 <= x < size:
                pixels[y * size + x] = color
    return pixels

def main():
    output_dir = os.path.join(os.path.dirname(__file__), "..", "images")
    os.makedirs(output_dir, exist_ok=True)

    icons = {
        "home": draw_house,
        "academic": draw_book,
        "resume": draw_document,
        "profile": draw_person,
    }

    normal_color = rgba(153, 153, 153, 255)   # #999
    active_color = rgba(74, 144, 217, 255)     # #4A90D9

    for name, draw_fn in icons.items():
        # Normal (gray)
        pixels_normal = draw_fn(normal_color)
        png_data = create_png(81, 81, pixels_normal)
        path = os.path.join(output_dir, f"{name}.png")
        with open(path, "wb") as f:
            f.write(png_data)
        print(f"Created: {path} ({len(png_data)} bytes)")

        # Active (blue)
        pixels_active = draw_fn(active_color)
        png_data = create_png(81, 81, pixels_active)
        path = os.path.join(output_dir, f"{name}-active.png")
        with open(path, "wb") as f:
            f.write(png_data)
        print(f"Created: {path} ({len(png_data)} bytes)")

if __name__ == "__main__":
    main()

File: frontend-miniapp/scripts/generate_icons.js
/** Generate tabBar icons for WeChat mini-program. No dependencies needed. */
const zlib = require("zlib");
const fs = require("fs");
const path = require("path");

function createPng(width, height, pixels) {
  function chunk(type, data) {
    const len = Buffer.alloc(4);
    len.writeUInt32BE(data.length);
    const typeAndData = Buffer.concat([Buffer.from(type, "ascii"), data]);
    const crc = Buffer.alloc(4);
    crc.writeUInt32BE(zlib.crc32(typeAndData));
    return Buffer.concat([len, typeAndData, crc]);
  }

  const sig = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);

  const ihdrData = Buffer.alloc(13);
  let off = 0;
  ihdrData.writeUInt32BE(width, off); off += 4;
  ihdrData.writeUInt32BE(height, off); off += 4;
  ihdrData[off++] = 8;  // bit depth
  ihdrData[off++] = 6;  // color type: RGBA
  ihdrData[off++] = 0;  // compression
  ihdrData[off++] = 0;  // filter
  ihdrData[off++] = 0;  // interlace

  // Build raw image data: filter byte 0 + RGBA for each row
  const rawRows = [];
  for (let y = 0; y < height; y++) {
    const row = Buffer.alloc(1 + width * 4);
    row[0] = 0; // filter: None
    for (let x = 0; x < width; x++) {
      const p = pixels[y * width + x];
      const base = 1 + x * 4;
      row[base] = p[0];
      row[base + 1] = p[1];
      row[base + 2] = p[2];
      row[base + 3] = p[3];
    }
    rawRows.push(row);
  }

  return Buffer.concat([
    sig,
    chunk("IHDR", ihdrData),
    chunk("IDAT", zlib.deflateSync(Buffer.concat(rawRows))),
    chunk("IEND", Buffer.alloc(0)),
  ]);
}

function transparent() { return [0, 0, 0, 0]; }
function color(r, g, b, a = 255) { return [r, g, b, a]; }

function drawIcon(drawFn, c, size = 81) {
  const pixels = [];
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      pixels.push(drawFn(x, y, size, c));
    }
  }
  return pixels;
}

function drawHouse(x, y, s, c) {
  // Roof triangle
  if (y >= 12 && y < 35) {
    const roofWidth = Math.floor((y - 12) * 50 / 23 + 10);
    if (x >= 40 - roofWidth && x <= 40 + roofWidth) return c;
  }
  // Body
  if (y >= 30 && y < 75 && x >= 15 && x < 66) {
    // Door
    if (y >= 48 && y < 75 && x >= 32 && x < 49) return color(255, 255, 255);
    return c;
  }
  return transparent();
}

function drawBook(x, y, s, c) {
  // Book body
  if (y >= 20 && y < 65 && x >= 20 && x < 61) {
    // Spine
    if (x >= 38 && x < 43) return color(255, 255, 255, 200);
    // Text lines
    if ([28, 36, 44, 52].includes(y) && ((x >= 24 && x < 36) || (x >= 45 && x < 57))) {
      return color(255, 255, 255, 150);
    }
    return c;
  }
  return transparent();
}

function drawDocument(x, y, s, c) {
  // Paper body
  if (y >= 15 && y < 70 && x >= 18 && x < 63) {
    // Fold corner
    if (y < 30 && x >= 48 && (x - 48) >= (30 - y)) return transparent();
    // Text lines
    if ([28, 36, 44, 52, 60].includes(y) && x >= 26 && x < 56) return color(255, 255, 255, 180);
    return c;
  }
  return transparent();
}

function drawPerson(x, y, s, c) {
  // Head (ellipse)
  const dx = x - 40.5, dy = y - 25;
  if (dx * dx / 169 + dy * dy / 169 <= 1) return c;
  // Body (trapezoid)
  if (y >= 40 && y < 72) {
    const bw = 8 + Math.floor((y - 40) * 6 / 16);
    if (x >= 40 - bw && x <= 40 + bw) return c;
  }
  return transparent();
}

const icons = {
  home: drawHouse,
  academic: drawBook,
  resume: drawDocument,
  profile: drawPerson,
};

const normalColor = color(153, 153, 153);  // #999
const activeColor = color(74, 144, 217);   // #4A90D9

const outputDir = path.join(__dirname, "..", "images");
fs.mkdirSync(outputDir, { recursive: true });

for (const [name, drawFn] of Object.entries(icons)) {
  for (const [suffix, clr] of [["", normalColor], ["-active", activeColor]]) {
    const pixels = drawIcon(drawFn, clr);
    const png = createPng(81, 81, pixels);
    const filePath = path.join(outputDir, `${name}${suffix}.png`);
    fs.writeFileSync(filePath, png);
    console.log(`Created: ${filePath} (${png.length} bytes)`);
  }
}
